entry_types/scrolled/package/src/frontend/MediaPlayer/index.js

```diff
@@ -0,0 +1,52 @@
+import React, {useEffect, useRef, useContext} from 'react';
+import {media} from 'pageflow/frontend';
+
+import ScrollToSectionContext from "../ScrollToSectionContext";
+
+export function MediaPlayer(props){
+  const playerWrapperRef = useRef(null);
+  let scrollToSection = useContext(ScrollToSectionContext);
+
+  useEffect( () => {
+    let playerWrapper = playerWrapperRef.current;
+    
+    if (props.sources) {
+      let player = media.getPlayer(props.sources, {
+        poster: props.poster,
+        tagName: props.type,
+        playsInline: props.playsInline,
+        loop: props.loop,
+        controls: props.controls
+      });
+      player.on('ended', function () {
+        props.nextSectionOnEnd && scrollToSection('next');
+      });
+      let playerElement = player.el();      
+      playerWrapper.appendChild(playerElement);
+      
+      if (!media.mediaOff && props.autoplay !== false) {
+        if (props.state === 'active') {
+          if (player.readyState() > 0) {
+            player.play();
+          } else {
+            player.on('loadedmetadata', player.play);
+          }
+        } else {
+          player.pause();
+        }
+      }
+
+      return () => {
+        media.releasePlayer(player);
+        playerWrapper.innerHTML = '';
+        player.dispose();
+      }
+    }
+  }, [scrollToSection, props.autoplay, props.state, props.sources, props.poster, props.type,
+      props.loop, props.controls, props.nextSectionOnEnd, props.playsInline]);
```

Comment:
Looking at how complicated it is to construct a memoized array of sources in the two components above, makes me think we should reconsider our approach here. Also while testing, I still saw unneeded recreation of the player when I selected elements in the editor and in some cases on scroll in the storybook.

The solution implemented here is exactly what we discussed, but looking at it now and testing it, I think the main problem is that we depend on clients of `MediaPlayer` component to pass stable props to ensure the player is not recreated. The reason the player is not recreated should be a feature of the `MediaPlayer` itself, though. Especially since `MediaPlayer` receives more props than is actually needed for the player creation (e.g. `nextSectionOnEnd`), this problem will otherwise always come up again whenever we add something to `MediaPlayer` or use it in a new setting.

To achieve this, I think we can use a similar approach to how `MediaTag` works in `pageflow-paged-react`, which uses `shouldComponentUpdate` to re-render the component only if props changed in a substantial way. The counterpart of `shouldComponentUpdate` for function components is [`React.memo`](https://reactjs.org/docs/react-api.html#reactmemo). I suggest to extract a component (which could be called `PlayerContainer`) that takes only the props required to call `getPlayer`. It recreates the player in a `useEffect` that runs on every render and then uses `React.memo` with a custom comparison function to only rerender when the props changed. This comparison can work exactly like in `pageflow-paged-react`'s `MediaTag` with a deep comparison of sources. That way it does not matter to pass the same instance of the `sources` object, but just one with equal urls.

It can handle `onSetup`/`onDispose` props just like `MediaTag` to pass the created `player` object out. The rest of the player state/autoplay logic can then continue to live in `MediaPlayer`. That way autoplay code will also live in a different effect than player creation. The above code recreated the player each time `props.state` changes.